java/ql/lib/semmle/code/java/frameworks/MyBatis.qll

```diff
@@ -105,94 +105,88 @@ class TypeParam extends Interface {
-          "SELECT", "OFFSET_ROWS", "FETCH_FIRST_ROWS_ONLY", "OFFSET", "LIMIT", "ORDER_BY", "HAVING",
-          "GROUP_BY", "WHERE", "OUTER_JOIN", "RIGHT_OUTER_JOIN", "LEFT_OUTER_JOIN", "INNER_JOIN",
-          "JOIN", "FROM", "DELETE_FROM", "SELECT_DISTINCT", "SELECT", "INTO_VALUES", "INTO_COLUMNS",
-          "VALUES", "INSERT_INTO", "SET", "UPDATE"
-        ]
-    }
+private class MyBatisProvider extends RefType {
+  MyBatisProvider() {
+    this.hasQualifiedName("org.apache.ibatis.annotations",
+      ["Select", "Delete", "Insert", "Update"] + "Provider")
   }
+}
 
-  class MyBatisInjectionSink extends DataFlow::Node {
-    MyBatisInjectionSink() {
-      exists(Annotation a, Method m, TypeLiteral type, Class c |
-        a.getType() instanceof MyBatisProvider and
-        type = a.getValue(["type", "value"]) and
-        c.hasMethod(m, type.getTypeName().getType()) and
-        m.hasName(a.getValue("method").(StringLiteral).getValue()) and
-        this.asExpr() = m.getBody().getAStmt().(ReturnStmt).getResult()
-      )
-    }
+private class MyBatisAbstractSQLMethodNames extends string {
+  MyBatisAbstractSQLMethodNames() {
+    this in [
+        "SELECT", "OFFSET_ROWS", "FETCH_FIRST_ROWS_ONLY", "OFFSET", "LIMIT", "ORDER_BY", "HAVING",
+        "GROUP_BY", "WHERE", "OUTER_JOIN", "RIGHT_OUTER_JOIN", "LEFT_OUTER_JOIN", "INNER_JOIN",
+        "JOIN", "FROM", "DELETE_FROM", "SELECT_DISTINCT", "SELECT", "INTO_VALUES", "INTO_COLUMNS",
+        "VALUES", "INSERT_INTO", "SET", "UPDATE"
+      ]
   }
+}
 
-  class MyBatisAdditionalTaintStep extends TaintTracking::AdditionalTaintStep {
-    abstract override predicate step(DataFlow::Node node1, DataFlow::Node node2);
+class MyBatisInjectionSink extends DataFlow::Node {
+  MyBatisInjectionSink() {
+    exists(Annotation a, Method m, TypeLiteral type, Class c |
+      a.getType() instanceof MyBatisProvider and
+      type = a.getValue(["type", "value"]) and
+      c.hasMethod(m, type.getTypeName().getType()) and
+      m.hasName(a.getValue("method").(StringLiteral).getValue()) and
+      this.asExpr() = m.getBody().getAStmt().(ReturnStmt).getResult()
+    )
   }
+}
 
-  private class MyBatisProviderStep extends MyBatisAdditionalTaintStep {
-    override predicate step(DataFlow::Node n1, DataFlow::Node n2) {
-      exists(
-        MethodAccess ma, Annotation a, Method annotatedMethod, Method providerMethod,
-        TypeLiteral type, Class c
-      |
-        a.getType() instanceof MyBatisProvider and
-        annotatedMethod.getAnAnnotation() = a and
-        ma.getMethod() = annotatedMethod and
-        ma.getAnArgument() = n1.asExpr() and
-        type = a.getValue(["type", "value"]) and
-        providerMethod.hasName(a.getValue("method").(StringLiteral).getValue()) and
-        c.hasMethod(providerMethod, type.getTypeName().getType()) and
-        providerMethod.getAParameter() = n2.asParameter()
-      )
-    }
+private class MyBatisProviderStep extends TaintTracking::AdditionalTaintStep {
```

Comment:
Now that I think of it, this is actually an `AdditionalValueStep`, since `n1` and `n2` are the same value.

Reply:
Done https://github.com/github/codeql/pull/8345/commits/8482c019595595e407f713e9f121f6514d336ce4